js/node_shared/unix.js

```diff
@@ -0,0 +1,369 @@
+  console.error("unix_chdir");
+  console.error(path);
+
+  caml_sys_chdir(path);
+
+  return;
+}
+
+//Provides: unix_clear_close_on_exec
+function unix_clear_close_on_exec(fd) {
+  console.error("unix_clear_close_on_exec");
+  console.error(fd);
+  return;
+}
+
+//Provides: unix_dup
+//Requires: unix_pipe
+function unix_dup(cloexec, fd) {
+  console.error("unix_dup");
+  console.error(cloexec);
+  console.error(fd);
+
+  // We just use the same unix_pipe logic to generate a fresh fake
+  // file descriptor.
+  return unix_pipe(cloexec, fd);
+}
+
+//Provides: unix_close
+function unix_close(fd) {
+  console.error("unix_close");
+  console.error(fd);
+
+  return;
+}
+
+//Provides: unix_fork
+function unix_fork(vunit) {
+  console.error("unix_fork");
+  console.error(vunit);
+
+  return;
+}
+
+//Provides: unix_setitimer
+function unix_setitimer(it, its) {
+  console.error("unix_setitimer");
+  console.error(it);
+  console.error(its);
+
+  // In our code, the result value is just ignored, so we can return the same one.
+  // TODO: Theoretically this is supposed to do some SIGALRM stuff too, but
+  // that is complicated, so let's leave it as a TODO.
+  return its;
+}
+
+// TODO
+//Provides: unix_sleep
+function unix_sleep(duration) {
+  console.error("unix_sleep");
+  console.error(duration);
+
+  // You cannot actually sleep synchronously, so let's just leave this
+  // as a TODO.
+
+  return;
+}
+
+// TODO: This is fixed as of this PR:
+// https://github.com/ocsigen/js_of_ocaml/pull/1519
+// so when this is released, we will just need to upgrade our jsoo version
+//Provides: caml_unix_lstat_64
+//Requires: caml_unix_lstat, caml_int64_of_int32
+//Alias: unix_lstat_64
+function caml_unix_lstat_64(name) {
+  var r = caml_unix_lstat(name);
+  r[9] = caml_int64_of_int32(r[9]);
+  return r;
+}
+
+//Provides: unix_realpath
+function unix_realpath(path) {
+  return globalThis.fs.realpathSync(path);
+}
+
+//Provides: unix_umask
+function unix_umask(newmask) {
+  let oldmask = globalThis.process.umask(newmask);
+  return oldmask;
+}
+
+//Provides: unix_write
+//Requires: caml_sys_fds
+function unix_write(fd, buf, ofs, len) {
+  console.error("unix_write");
+  console.error(fd);
+  console.error(ofs);
+  console.error(len);
+
+  let file = caml_sys_fds[fd];
+
+  // The reason why this gymnastics is needed, instead of just calling file.write
+  // on the buffer buf.c, is because for some reason the call to writeSync is
+  // copying all of the bytes of the buffer, instead of the requested length from
+  // the offset.
+  // I observed previously that, for a 1024 byte buffer, it would copy all 1024
+  // bytes to the target, instead of the requested 178.
+  // So we just construct a new buffer of what we really want, to ensure the call
+  // gets it right.
+  let bufferSlice = buf.c.slice(ofs, ofs + len);
+
+  // The actual length of our buffer, which depends upon whether the index of
+  // ofs + len is greater than the buffer that we are reading from.
```

Comment:
Can you explain this a bit more? I.e. What happens if ofs + len > bufferSlice.length, or why that's impossible. 

I wonder if you can just return the length of the slice right instead of doing some conditional stuff here? won't it always be len

Reply:
The length shouldn't always be len I think, depending on if the len + ofs asked for is actually greater than the bounds of the array.

That being said, I think you're correct that this just ends up being the length of the buffer. I'll change that.